File: express/src/routes/Marketplace/trade_bids.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { TradeBidService } from '../../services/Marketplace/trade_bid_service.js';

const router = Router();
const service = new TradeBidService();

function validate(data: any): void {
  if (!((data.amount == null || Number(data.amount) > 0))) throw new Error(`Bid amount must be greater than zero`);
}
function applyProjection(obj: any): any {
  if (!obj) return obj;
  const r = { ...obj };
  if ('placedAt' in r) { r.placedAt = r.placedAt; delete r.placedAt; }
  return r;
}


router.get('/', async (req, res) => {
  const items = await prisma.tradeBid.findMany();
  res.json(items.map(applyProjection));
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.amount !== undefined) data.amount = body.amount;
    if (body.placedAt !== undefined) data.placedAt = body.placedAt != null ? new Date(body.placedAt) : null;
    if (body.isWinning !== undefined) data.isWinning = body.isWinning;
    if (body.listingId !== undefined) data.listingId = body.listingId;
    if (body.bidderId !== undefined) data.bidderId = body.bidderId;
  try {
  validate(data);
    const entity = await prisma.tradeBid.create({ data });
    res.status(201).json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Validation error') });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.tradeBid.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(applyProjection(entity));
});

router.get('/:id/outbid', async (req, res) => {
  const id = Number((req.params as any).id);
  const newAmount = (req.query as any).newAmount;
  try {
    const result = await service.outbid_by(id, newAmount);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.delete('/:id', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.retract(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});
export default router;

File: gin/internal/handler/marketplace/trade_bids.go
package handler_marketplace

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/marketplace"
	"cards_project/internal/handler"
)

type TradeBidHandler struct { db *gorm.DB }

func NewTradeBidHandler(db *gorm.DB) *TradeBidHandler {
	return &TradeBidHandler{db: db}
}

func (h *TradeBidHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/trade_bids")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.GET("/:id/api/bids/{id}/outbid", h.OutbidBy)
	g.DELETE("/:id/api/bids/{id}", h.Retract)
}

func (h *TradeBidHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.TradeBid
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.TradeBidResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *TradeBidHandler) Create(c *gin.Context) {
	var req model.TradeBidCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validateTradeBid(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.TradeBid{}
	row.Amount = req.Amount
	row.PlacedAt = req.PlacedAt
	row.IsWinning = req.IsWinning
	row.ListingID = req.ListingID
	row.BidderID = req.BidderID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *TradeBidHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeBid
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeBid"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *TradeBidHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeBid
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeBid"); return }
		handler.DbError(c, err); return
	}
	var req model.TradeBidUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestTradeBid(&row)
	if msgs := validateTradeBid(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *TradeBidHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *TradeBidHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.TradeBid{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeBid"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *TradeBidHandler) OutbidBy(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeBid
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeBid"); return }
		handler.DbError(c, err); return
	}
	var body map[string]interface{}
	_ = c.ShouldBindJSON(&body)
	newAmount := func() float64 {
		v, ok := body["new_amount"]; if !ok { return 0 }
		f, ok := v.(float64); if !ok { return 0 }
		return f
	}()
	result, err := row.OutbidBy(newAmount)
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func (h *TradeBidHandler) Retract(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.TradeBid
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "TradeBid"); return }
		handler.DbError(c, err); return
	}
	err := row.Retract()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func validateTradeBid(req *model.TradeBidCreateRequest) []string {
	var errs []string
	if !(float64(req.Amount) > 0) {
		errs = append(errs, "Bid amount must be greater than zero")
	}
	return errs
}

func toCreateRequestTradeBid(m *model.TradeBid) model.TradeBidCreateRequest {
	return model.TradeBidCreateRequest{
		Amount: m.Amount,
		PlacedAt: m.PlacedAt,
		IsWinning: m.IsWinning,
		ListingID: m.ListingID,
		BidderID: m.BidderID,
	}
}
